File: src/cpp02_service/src/demo01_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "base_interfaces/srv/add_two_ints.hpp"

using base_interfaces::srv::AddTwoInts;
using std::placeholders::_1;
using std::placeholders::_2;

class MyServer: public rclcpp::Node 
{
public:
  MyServer(std::string name)
  : Node(name)
  , cnt(0)
  {
    AddServer = this->create_service<AddTwoInts>(
      "add_server",
      std::bind(&MyServer::add_callback,this,_1,_2)
    );
  }

private:
  int cnt;
  rclcpp::Service<AddTwoInts>::SharedPtr AddServer;

  void add_callback(const AddTwoInts::Request::SharedPtr req, 
                    AddTwoInts::Response::SharedPtr res) const
  {
    res->sum = req->num1 + req->num2;
    RCLCPP_INFO(this->get_logger(),"sum=%d\n",res->sum);
  }
};

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  
  auto node = std::make_shared<MyServer>("server_node");
  rclcpp::spin(node);
  
  rclcpp::shutdown();
  
  return 0;
}

File: ws_zxz_chapt2/src/py02_service/py02_service/demo01_server.py
import rclpy
from rclpy.node import Node
from base_interfaces.srv import AddTwoInts

class MinimalService(Node):

    def __init__(self):
        super().__init__('minimal_service_py')
        self.srv = self.create_service(AddTwoInts, 'add_ints', self.add_two_ints_callback)
        self.get_logger().info("服务端启动！")

    def add_two_ints_callback(self, request, response):
        response.sum = request.num1 + request.num2
        self.get_logger().info('请求数据:(%d,%d),响应结果:%d' % (request.num1, request.num2, response.sum))
        return response


def main():
    rclpy.init()

    minimal_service = MinimalService()
    rclpy.spin(minimal_service)

    rclpy.shutdown()


if __name__ == '__main__':
    main()